### PR TITLE
LNW v2 user guide.

### PR DESCRIPTION
**Pull request by n-sandeep**

### CODE REVIEW

**Comment by n-sandeep** [comment]
This PR will be appended with usecase documentation. Which will replace content of
es2k-linux-networking.md
es2k-linux-networking-ecmp.md
es2k-linux-networking-ipv6.md
es2k-linux-networking-frr.md
es2k-linux-networking-lag.md

File: docs/apps/lnw/es2k/readme.md [hunk 1]
```diff
@@ -0,0 +1,273 @@
+<!--
```
Comment (5abeel) [inline docs/apps/lnw/es2k/readme.md:1]: Dont think we need a license block for a README page

File: docs/apps/lnw/es2k/readme.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+# Linux Networking for ES2K
+
+Linux Networking involves incorporating comprehensive support for offloading various networking functions, such as L2 forwarding, L3 forwarding, ECMP, and VXLAN encapsulation and decapsulation intelligence to the IPU. This capability empowers overlay services to establish communication with endpoints through VxLAN tunnels, thereby extending the L2 segment across the underlay network. To achieve Linux networking support, we have utilized legacy OVS for overlay source MAC learning and VxLAN configurations, while relying on the kernel for underlay neighbor discovery, route management, and next-hop information.
+
+## Feature Overview
+
+This feature can run in two modes:
+
+- Slow path: All packets must land on the control plane and only control plane has the intelligence to forward the traffic.
+- Fast path: Onlt 1st unknown traffic (both SMAC and DMAC) will be sent to control plane and subsequent traffic should be forwarded by IPU E2100.
```
Comment (5abeel) [inline docs/apps/lnw/es2k/readme.md:30]: Typo
```suggestion
- Fast path: Only 1st unknown traffic (both SMAC and DMAC) will be sent to control plane and subsequent traffic should be forwarded by IPU E2100.
```

File: docs/apps/lnw/es2k/readme.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+To enable this feature we have,
+- `Infrap4d`: This process has p4runtime intelligence and acts as a gRPC server. Calls TDI front end to program IPU E2100.
+- `ovs-vswitchd`: This process is integrated with p4runtime intelligence and acts as a gRPC client. Programs IPU E2100 with control plane configuration and forwrding tables by communicating with gRPC server.
```
Comment (5abeel) [inline docs/apps/lnw/es2k/readme.md:34]: Typo

```suggestion
- `ovs-vswitchd`: This process is integrated with p4runtime intelligence and acts as a gRPC client. Programs IPU E2100 with control plane configuration and forwarding tables by communicating with gRPC server.
```

File: docs/apps/lnw/es2k/readme.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+- `p4rt-ctl`: This python CLI has p4runtime intelligence and acts as a gRPC client. Programs IPU E2100 with runtime rules by communicating with gRPC server.
+- `Kernel stack`: All underlay related configurations are picked by `kernel monitor` thread via netlink events in `infrap4d` and these are programmed in IPU E2100 by calling TDI front end calls.
+
+## Topology
+
+This topology breakup and configuration assumes, all VMs are spawned on HOST VFs and control plane is running in ACC.
```
Comment (5abeel) [inline docs/apps/lnw/es2k/readme.md:40]:
```suggestion
This topology breakup and configuration assumes all VMs are spawned on HOST VFs and control plane is running in ACC.
```

No need for comma after 'assumes'

File: docs/apps/lnw/es2k/readme.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+![Linux Networking Topology](es2k-lnw-topology.png)
+
+### Topology breakup
+
+- Every VM spawned on top of a VF will have a corresponding port representer in ACC.
+- Every physical port will have a corresponding port representer in ACC.
+- Every physical port will have an uplink (APF netdev) in HOST and this uplink will have a corresponding port representer in ACC.
+- All port representers are associated with an OvS bridge.
+- For VxLAN egress traffic, underlay port should be associated with a termination bridge and IP ro reach underlay network should be configured on top of this bridge.
```
Comment (satish153) [inline docs/apps/lnw/es2k/readme.md:50]: Type: "ro reach"=> "to reach"

File: docs/apps/lnw/es2k/readme.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+## Detailed Design
+
+### Slow path
+
+To enable slow path mode,
+- Start infrap4d process by disabling `kernel monitor` thread. Command: `infrap4d -disable-krnlmon`
+- Set environment variable `OVS_P4_OFFLOAD=false` before starting `ovs-vswitchd` process.
+
+In this mode, we need to associate VFs on top which VMs are created and its port representers and als physical ports with its port representers.
```
Comment (satish153) [inline docs/apps/lnw/es2k/readme.md:60]: Typo: als => also

File: docs/apps/lnw/es2k/readme.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+Once these tables are configured refer to packet flow as mentioned below.
+
+#### For TX:
+
+##### East traffic without VxLAN encap
+
+Packets coming from overlay network,
+- Determine the Source port of the packet based on which overlay VSI the packet has landed.
+- Validate if the source port is part of the bridge, else drop the packet.
+- If valid bridge configuration is found, find the PR assocaited with the bridge and forward the packet to the PR in ACC.
```
Comment (satish153) [inline docs/apps/lnw/es2k/readme.md:87]: associated

File: docs/apps/lnw/es2k/readme.md [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+For egress VxLAN traffic, an OvS VxLAN port need to be created in ACC and associated to the integration bridge that handles overlay traffic.
+Configure table:
+- rx_ipv4_tunnel_source_port/rx_ipv6_tunnel_source_port
+- source_port_to_bridge_map
+
+Once these tables are configured refer to packet flow as mentioned below.
+
+#### For TX:
+
+##### East traffic without VxLAN encap
```
Comment (satish153) [inline docs/apps/lnw/es2k/readme.md:82]: Egress traffic

File: docs/apps/lnw/es2k/readme.md [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+This topology breakup and configuration assumes, all VMs are spawned on HOST VFs and control plane is running in ACC.
+
+![Linux Networking Topology](es2k-lnw-topology.png)
+
+### Topology breakup
+
+- Every VM spawned on top of a VF will have a corresponding port representer in ACC.
+- Every physical port will have a corresponding port representer in ACC.
+- Every physical port will have an uplink (APF netdev) in HOST and this uplink will have a corresponding port representer in ACC.
```
[the remark this answers is not in the document]
Comment (5abeel) [reply docs/apps/lnw/es2k/readme.md]: Good catch. Fixing it

File: docs/apps/lnw/es2k/readme.md (hunk 1 above)
Comment (5abeel) [reply docs/apps/lnw/es2k/readme.md]: Removed it

**Review by 5abeel — APPROVED** [review]

File: docs/apps/lnw/es2k/readme.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+## Detailed Design
+
+### Slow path
+
+To enable slow path mode:
+- Start infrap4d process by disabling `kernel monitor` thread. Command: `infrap4d -disable-krnlmon`
+- Set environment variable `OVS_P4_OFFLOAD=false` before starting `ovs-vswitchd` process.
+
+In this mode, we need to associate VFs on top which VMs are created and its port representors and also physical ports with its port representors.
```
Comment (ffoulkes) [inline docs/apps/lnw/es2k/readme.md:41]:
```suggestion
In this mode, we need to associate VFs on top of which VMs are created and its port representors and also physical ports with its port representors.
```
This sentence is garbled. I can't figure out the relationships between VFs, VMs, physical ports, and port representors.

You're using _its_ (singular possessive) with plural nouns (_VFs_, _VMs_, _ports_, _representors_). The plural possessive is _their_. "ports with _their_ representors" (assuming that's the intended pairing).

File: docs/apps/lnw/es2k/readme.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+- Every physical port will have an uplink (APF netdev) in HOST and this uplink will have a corresponding port representor in ACC.
+- All port representors are associated with an OvS bridge.
+- For VxLAN egress traffic, underlay port should be associated with a termination bridge and IP to reach underlay network should be configured on top of this bridge.
+
+## Detailed Design
+
+### Slow path
+
+To enable slow path mode:
+- Start infrap4d process by disabling `kernel monitor` thread. Command: `infrap4d -disable-krnlmon`
```
Comment (ffoulkes) [inline docs/apps/lnw/es2k/readme.md:38]:
```suggestion
- Start the infrap4d process with the Kernel Monitor disabled. Command: `infrap4d -disable-krnlmon`
```

**Review by ffoulkes — COMMENTED** [review]
Only a partial review. I'll try to get back to it later.

File: docs/apps/lnw/es2k/readme.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+
+## Feature Overview
+
+This feature can run in two modes:
+
+- Slow path: All packets must land on the control plane and only control plane has the intelligence to forward the traffic.
+- Fast path: Only first unknown traffic (both SMAC and DMAC) will be sent to control plane and subsequent traffic should be forwarded by IPU E2100.
+
+To enable this feature we have,
+- `Infrap4d`: This process has p4runtime intelligence and acts as a gRPC server. Calls TDI front end to program IPU E2100.
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply docs/apps/lnw/es2k/readme.md:14]: We cannot modify this as "p4runtime server", as its aGRPC server which uses p4runtime format.

File: docs/apps/lnw/es2k/readme.md [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+- `ovs-vswitchd`: This process is integrated with p4runtime intelligence and acts as a gRPC client. Programs IPU E2100 with control plane configuration and forwarding tables by communicating with gRPC server.
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply docs/apps/lnw/es2k/readme.md:15]: same as above

File: docs/apps/lnw/es2k/readme.md [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+- `p4rt-ctl`: This python CLI has p4runtime intelligence and acts as a gRPC client. Programs IPU E2100 with runtime rules by communicating with gRPC server.
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply docs/apps/lnw/es2k/readme.md:16]: same as above

File: docs/apps/lnw/es2k/readme.md [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+- Verification of Source Port and Associated L2 Bridge: The P4 Control Plane (P4 CP) must ensure the validation of the source port and its corresponding L2 bridge before initiating any further regulation of datapath packet classification.
+- Exception Packet Handling for All Protocols: The P4 Control Plane (P4 CP) shall incorporate exception packet handling logic, not limited to ARP but applicable to the first packet of any protocol.
+- Offloading of Networking Functions: The P4 Control Plane (P4 CP) software shall provide support for the offloading of various networking functions as specified in the Linux Networking use case. These networking functions include Layer 2 (L2) and Layer 3 (L3) forwarding, Equal-Cost Multi-Path (ECMP) routing, Link Aggregation Group (LAG), as well as Virtual Extensible LAN (VXLAN) encapsulation and decapsulation. These functions shall support both single and multiple Open vSwitch (OvS) bridges.
+
+
+## Limitations
+
+Current Linux Networking support for the networking recipe has following limitations:
+
+- VLAN configuration on OvS is supported only for NATIVE-TAG and NATIVE-UNAG modes.
```
[the remark this answers is not in the document]
Comment (n-sandeep) [reply docs/apps/lnw/es2k/readme.md:257]: ahh , its NATIVE-UNTAG

File: docs/apps/lnw/es2k/readme.md
[the remark this answers is not in the document]
Comment (n-sandeep) [reply docs/apps/lnw/es2k/readme.md:1]: Sure.. let me download markdownlint and fix these. Thanks Derek

File: docs/apps/lnw/es2k/readme.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ffoulkes) [reply docs/apps/lnw/es2k/readme.md]: We generally speak of a server in terms of the services it provides, not the transport protocols used. We say "FTP server" or "HTTP server", not "TCP server."

The gRPC services that Stratum supports are P4runtime (p4_service), gNMI (config_monitoring_service), and gNOI (admin_service, certificate_management_service, diag_service, file_service). We don't enable gNOI in our targets, so infrap4d provides a P4Runtime server and a gNMI server.

[the remark this answers is not in the document]
Comment (n-sandeep) [reply docs/apps/lnw/es2k/readme.md]: Got it Derek, let me modify as suggested.

**Review by ffoulkes — APPROVED** [review]
There's no point in my investing time any more time reviewing this document if you're going to ignore my comments.

Approved without further comment.

**Pull request merged**